sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -1174,7 +1181,136 @@ case class GpuRegExpExtract(
+  extends GpuRegExpTernaryBase with ImplicitCastInputTypes with NullIntolerant {
+
+  override def dataType: DataType = ArrayType(StringType, containsNull = true)
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType, IntegerType)
+  override def first: Expression = str
+  override def second: Expression = regexp
+  override def third: Expression = idx
+
+  override def prettyName: String = "regexp_extract_all"
 
+  override def doColumnar(
+      str: GpuColumnVector,
+      regexp: GpuScalar,
+      idx: GpuScalar): ColumnVector = {
+
+    idx.getValue.asInstanceOf[Int] match {
+      case 0 => 
+        str.getBase.extractAllRecord(cudfRegexPattern, 0)
+      case intIdx =>
+        // Extract matches corresponding to idx. cuDF's extract_all_record does not support 
+        // group idx, so we must manually extract the relevant matches. Example:
+        // Given the pattern (\d+)-(\d+) and idx=1
+        //
+        // |      Input      |      Java       |               cuDF             |
+        // |-----------------|-----------------|--------------------------------|
+        // | '1-2, 3-4, 5-6' | ['1', '3', '5'] | ['1', '2', '3', '4', '5', '6'] |
+        //
+        // Since idx=1 and the pattern has 2 capture groups, we take the 1st element and every
+        // 2nd element afterwards from the cuDF list
+
+        val rowCount = str.getRowCount
+        
+        val extractedWithNulls = withResource(
+          str.getBase.extractAllRecord(cudfRegexPattern, intIdx)) { allExtracted =>
+            withResource(allExtracted.countElements) { listSizes =>
+              withResource(listSizes.max) { maxSize =>
+                val maxSizeInt = maxSize.getInt
+                val stringCols: Seq[ColumnVector] = Range(intIdx - 1, maxSizeInt, numGroups).map { 
+                  i =>
+                    withResource(ColumnVector.fromScalar(Scalar.fromInt(i), rowCount.toInt)) { 
+                      index => allExtracted.extractListElement(index)
+                    }
+                }
+                ColumnVector.makeList(stringCols: _*)
+              }
+            }
+          }
+        // Filter out null values in the lists
+        val extractedStrings = withResource(extractedWithNulls.getListOffsetsView) { offsetsCol =>
+            withResource(extractedWithNulls.getChildColumnView(0)) { stringCol =>
+              withResource(stringCol.isNotNull) { isNotNull =>
+                withResource(isNotNull.makeListFromOffsets(rowCount, offsetsCol)) { booleanMask => 
+                  extractedWithNulls.applyBooleanMask(booleanMask)
+                }
+              }
+            }
+          }
+        // If input is null, output should also be null
+        withResource(GpuScalar.from(null, DataTypes.createArrayType(DataTypes.StringType))) { 
+          nullStringList =>
+            withResource(str.getBase.isNull) { isInputNull =>
+              withResource(extractedStrings) {
```

Comment:
We should have `withResource(extractedStrings)` as the first line in this block otherwise we could leak `extractedStrings` if there is an exception in one of the lines prior to this